fili-core/src/main/java/com/yahoo/bard/webservice/web/filterparser/ApiFiltersListListener.java

```diff
@@ -0,0 +1,219 @@
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_DIMENSION_UNDEFINED;
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_FIELD_NOT_IN_DIMENSIONS;
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.FILTER_OPERATOR_INVALID;
+
+/**
+ * Listener used to walk the Filters parse tree and build ApiFilter objects.
+ */
+public class ApiFiltersListListener extends FiltersBaseListener {
+    private static final Logger LOG = LoggerFactory.getLogger(ApiFiltersListListener.class);
+
+    private final Map<FiltersParser.FilterComponentContext, ApiFilter> filters = new HashMap<> ();
+    private final Map<Dimension, Set<ApiFilter>> dimensionFiltersMap = new LinkedHashMap<>();
+    private final List<Exception> errors = new LinkedList<>();
+    private final LogicalTable table;
+    private final DimensionDictionary dimensionDictionary;
+
+    /**
+     * Constructor.
+     *
+     * @param table logical table that the filter applies to
+     * @param dimensionDictionary dictionary of all the valid dimensions
+     */
+    public ApiFiltersListListener(LogicalTable table, DimensionDictionary dimensionDictionary) {
+        this.table = table;
+        this.dimensionDictionary = dimensionDictionary;
+    }
+
+    @Override
+    public void exitFilters(@NotNull FiltersParser.FiltersContext ctx) {
+        for (FiltersParser.FilterComponentContext f : ctx.filterComponent()) {
+            ApiFilter filter = filters.get(f);
+            if (filter == null) {
+                // there was some error and it has been saved in the errors list
+                break;
+            }
+            Dimension dimension = filter.getDimension();
+            if (!dimensionFiltersMap.containsKey(dimension)) {
+                dimensionFiltersMap.put(dimension, new LinkedHashSet<>());
+            }
+            Set<ApiFilter> filterSet = dimensionFiltersMap.get(dimension);
+            filterSet.add(filter);
+            dimensionFiltersMap.put(dimension, filterSet);
+        }
+    }
+
+    @Override
+    public void exitFilterComponent(@NotNull FiltersParser.FilterComponentContext f) {
+        Dimension dimension = extractDimension(f);
+        if (dimension == null) {
+            return;
+        }
+
+        DimensionField dimensionField = extractField(f, dimension);
+        if (dimensionField == null) {
+            return;
+        }
+
+        FilterOperation operation = extractOperation(f);
+        if (operation == null) {
+            return;
+        }
+
+        Set<String> values = extractValues(f);
+
+        ApiFilter filter = new ApiFilter(dimension, dimensionField, operation, values);
+
+        filters.put(f, filter);
+    }
+
+    /**
+     * Extract values from the filter.
+     * @param f The filter context
+     * @return the list of values
+     */
+    protected Set<String> extractValues(FiltersParser.FilterComponentContext f) {
```

Comment:
A better variable name than `f` would be really nice. This applies to the other methods of this style in this class as well.